rust/xaynet-mobile/src/ffi/participant.rs

```diff
@@ -0,0 +1,355 @@
+///    - The pointer must be properly [aligned].
+///    - It must be "dereferencable" in the sense defined in the
+///      [`::std::ptr`] module documentation.
+/// 2. After destroying the `Participant`, the pointer becomes invalid
+///    and must not be used.
+/// 3. This function should only be called on a pointer that has been
+///    created by [`xaynet_ffi_participant_new()`] or
+///    [`xaynet_ffi_participant_restore()`]
+///
+/// [`::std::ptr`]: https://doc.rust-lang.org/std/ptr/index.html#safety
+/// [aligned]: https://doc.rust-lang.org/std/ptr/index.html#alignment
+#[no_mangle]
+pub unsafe extern "C" fn xaynet_ffi_participant_destroy(participant: *mut Participant) -> c_int {
+    if participant.is_null() {
+        return ERR_NULLPTR;
+    }
+    pv::_xaynet_ffi_participant_destroy(participant);
+    OK
+}
+
+/// The participant is not taking part in the sum or update task
+pub const PARTICIPANT_TASK_NONE: c_int = 1;
+/// The participant is not taking part in the sum task
+pub const PARTICIPANT_TASK_SUM: c_int = 1 << 1;
+/// The participant is not taking part in the update task
+pub const PARTICIPANT_TASK_UPDATE: c_int = 1 << 2;
+/// The participant is expected to set the model it trained
+pub const PARTICIPANT_SHOULD_SET_MODEL: c_int = 1 << 3;
+/// The participant is expected to set the model it trained
+pub const PARTICIPANT_MADE_PROGRESS: c_int = 1 << 4;
+
+/// Instantiate a new participant with the given settings. The
+/// participant must be destroyed with [`xaynet_ffi_participant_destroy`].
+///
+/// # Return value
+///
+/// - a NULL pointer if `settings` is NULL or if the participant creation failed
+/// - a valid pointer to a [`Participant`] otherwise
+///
+/// # Safety
+///
+/// When calling this method, you have to ensure that *either* the
+/// pointer is NULL *or* all of the following is true:
+///
+/// - The pointer must be properly [aligned].
+/// - It must be "dereferencable" in the sense defined in the
+///   [`::std::ptr`] module documentation.
+///
+/// After destroying the participant with
+/// [`xaynet_ffi_participant_destroy`] becomes invalid and must not be
+/// used.
+///
+/// [`::std::ptr`]: https://doc.rust-lang.org/std/ptr/index.html#safety
+/// [aligned]: https://doc.rust-lang.org/std/ptr/index.html#alignment
+#[no_mangle]
+pub unsafe extern "C" fn xaynet_ffi_participant_new(settings: *const Settings) -> *mut Participant {
+    let settings = match unsafe { settings.as_ref() } {
+        Some(settings) => settings.clone(),
+        None => return std::ptr::null_mut(),
+    };
+
+    match Participant::new(settings) {
+        Ok(participant) => Box::into_raw(Box::new(participant)),
+        Err(_) => std::ptr::null_mut(),
+    }
+}
+
+/// Drive the participant internal state machine. Every tick, the
+/// state machine tries attempts to perform a small work unit.
+///
+/// # Return value
+///
+/// - [`ERR_NULLPTR`] is `participant` is NULL
+/// - a bitflag otherwise, with the following flags:
```

Comment:
Yeah it seems relatively common in C, since you can't return multiple values. We could also just return a struct, but since we're returning `int` in most places already a bitflag seemed appropriate